mode/cf/rest_client_test.go

```diff
@@ -1,13 +1,171 @@
+	port     = 8080
+	user     = "testuser"
+	pass     = "testpass"
+	pathElt1 = "path1"
+	pathElt2 = "path2"
+	pathElt3 = "path3"
+)
+
+func strSliceEq(s1, s2 []string) bool {
+	if len(s1) != len(s2) {
+		return false
+	}
+	for i, elt := range s1 {
+		if elt != s2[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func urlValuesEq(v1, v2 url.Values) bool {
+	lengths := len(v1) == len(v2)
+	if !lengths {
+		return false
+	}
+	for key, vals := range v1 {
+		if len(vals) != len(v2[key]) || !strSliceEq(vals, v2[key]) {
+			return false
+		}
+	}
+	return true
+}
+
+var (
+	queryStr = url.Values(map[string][]string{
+		"key1": []string{"val1"},
+		"key2": []string{"val2"},
+	})
 )
 
 func TestURLStr(t *testing.T) {
-	t.Skip("TODO")
+	rc := newRESTClient(http.DefaultClient, scheme, host, port, user, pass)
+	urlStr := rc.urlStr(pathElt1, pathElt2, pathElt3)
+	assert.Equal(t, urlStr, fmt.Sprintf("%s/%s/%s/%s", rc.fullBaseURL(), pathElt1, pathElt2, pathElt3), "url string")
 }
 
 func TestFullBaseURL(t *testing.T) {
-	t.Skip("TODO")
+	rc := newRESTClient(http.DefaultClient, scheme, host, port, user, pass)
+	fullURL := rc.fullBaseURL()
+	assert.Equal(t, fullURL, fmt.Sprintf("%s://%s:%s@%s:%d", scheme, user, pass, host, port), "full url string")
+}
+
+func testReq(
+	r *http.Request,
+	method,
+	scheme,
+	host string,
+	port int,
+	user,
+	pass string,
+	queryStr url.Values,
+	pathElts ...string,
+) error {
+	if r.Method != method {
+		return fmt.Errorf("method %s wasn't expected %s", r.Method, method)
+	}
+	if r.URL.Scheme != scheme {
+		return fmt.Errorf("scheme %s wasn't expected %s", r.URL.Scheme, scheme)
+	}
+	hostSpl := strings.Split(r.URL.Host, ":")
+	if len(hostSpl) != 2 {
+		return fmt.Errorf("invalid host string %s", r.URL.Host)
+	}
+	actualHost := hostSpl[0]
+	actualPort, err := strconv.Atoi(hostSpl[1])
+	if err != nil {
+		return fmt.Errorf("invalid port %s (%s)", hostSpl[1], err)
+	}
+	if actualHost != host {
+		return fmt.Errorf("host %s wasn't expected %s", actualHost, host)
+	}
+	if actualPort != port {
+		return fmt.Errorf("port %d wasn't expected %d", actualPort, port)
+	}
+	if user != r.URL.User.Username() {
+		return fmt.Errorf("username %s wasn't expected %s", r.URL.User.Username(), user)
+	}
+	realPass, _ := r.URL.User.Password()
+	if pass != realPass {
+		return fmt.Errorf("password %s wasn't expected %s", realPass, pass)
+	}
+	if !urlValuesEq(r.URL.Query(), queryStr) {
+		return fmt.Errorf("query string %s wasn't expected %s", r.URL.Query(), queryStr)
+	}
+	expectedPath := "/" + strings.Join(pathElts, "/")
+	if r.URL.Path != expectedPath {
+		return fmt.Errorf("path %s wasn't expected %s", r.URL.Path, expectedPath)
+	}
+	return nil
+}
+
+func TestGet(t *testing.T) {
+	rc := newRESTClient(http.DefaultClient, scheme, host, port, user, pass)
+	getReq, err := rc.Get(queryStr, pathElt1, pathElt2, pathElt3)
+	assert.NoErr(t, err)
+	assert.NoErr(t, testReq(getReq, "GET", scheme, host, port, user, pass, queryStr, pathElt1, pathElt2, pathElt3))
```

Comment:
If you passed `t` into `testReq(...)`, you could refactor `testReq(...)` to fail the tests itself instead of returning an error. It seems to me that the major advantage of that would be you could make more frequent use of functions in your `assert` package. Would that maybe even obviate the need for some of the url comparison logic you've implemented? (iirc, your `assert.Equal(...)` uses a `reflect.DeepEqual(...)` under the hood.